### ISSUE TITLE
Nfc error when approaching a yubico 5 nfc

### ISSUE THREAD

**Issue by tigernero79** [open-keychain/open-keychain#2650]
after updating to version 5.6 the encryption via NFC no longer works.  in practice, before I approached yubico Nfc to the cell phone and confirmation of encryption came out, now instead when it asks to approach the yubico via Nfc, it crashes or opens Chrome.  the control no longer applies at the time of the nfc

![Screenshot_20210215-184004](https://user-images.githubusercontent.com/35575744/107978627-693c6580-6fbd-11eb-849e-86a7fc970573.png)

![Screenshot_20210215-184028](https://user-images.githubusercontent.com/35575744/107978643-70637380-6fbd-11eb-940d-04e0df10ad2c.png)

**Comment by elijahmathews** [comment]
This is also happening to me. Usually, I would run into #2590 since I use an `ed25519` key, but the decryption would *eventually* work given enough tries with my Yubikey 5C NFC. However, the update to 5.6 seems to have broken this completely, as it just keeps failing even after dozens of attempts.

It also doesn't appear to just be an NFC issue, as I couldn't get my 5C to work when directly connected via USB-C either (which also worked before, eventually).

**Comment by tigernero79** [comment]
perfect then it is not only NFC problem also via type-c.  I kindly ask for an urgent update if possible

**Comment by a4b8** [comment]
Same issue on RSA keys. Therefore seems no correlation to cryptography key

**Comment by tigernero79** [comment]
then i hope an update to this issue will come out soon, otherwise application is unusable.

**Comment by sebdehne** [comment]
Same issuer here. Using RSA keys

**Comment by a4b8** [comment]
Just tested with a yubikey neo without any issues. Seems to be an issue regarding ybikey 5 only?

**Comment by sebdehne** [comment]
It works with version 5.5 but not with 5.6. Using YubiKey 5C NFC

**Comment by tigernero79** [comment]
@Valodim 

we all have the same problem, even with k9 mail v.5.730 when sending email interfacing with openpgp it fails to encrypt, opens yubico web page to opt

**Comment by hagau** [comment]
Hm, i've tested against an Yubikey 5 NFC with versions 5.1.2 & 5.2.7, with RSA & ECDSA keys, respectively.
No problems encrypting & decrypting mail with k9mail or text to/from clipboard.
(Mostly. Sometimes with ECDSA, bouncycastle throws an exception regarding the generated signature, but rather rarely. I'm pretty sure it's unrelated to this issue)

I'll try a Ed25519 key next, just to be sure, when I find the time.

Maybe this is an issue with Yubikey 5C NFC only?

**Comment by a4b8** [comment]
@hagau it's not an issue of 5C NFC. Working with a 5 NFC and the issue occurs. No matter with which other app I tested. Even if storing the gpg mime package in file system an trying to decrypt with open keychain v. 5.6 the open keychain app crashes. Including v. 5.5 everything was fine from my point of view. The update to 5.6 delivered 2 days broke the thing.

**Comment by a4b8** [comment]
@hagau which app version are you referencing with 5.1.2 / 5.2.7 
The problem occured with update to 5.6

**Comment by hagau** [comment]
@a4b8 those are Yubikey firmware version

**Comment by Valodim** [comment]
ugh, shit. sorry. we'll look into it shortly.

@hagau the KDF PR is one of the few changes in 5.6 that relate to smartcard usage. any chance it could be that?

**Comment by hagau** [comment]
@Valodim Could be :)

Unfortunately, I can't reproduce the issue.
I'll audit the code again later today/tomorrow.

**Comment by Valodim** [comment]
found the (likely) crash report on the play store:

```
java.lang.RuntimeException: 
  at android.os.AsyncTask$4.done (AsyncTask.java:415)
  at java.util.concurrent.FutureTask.finishCompletion (FutureTask.java:383)
  at java.util.concurrent.FutureTask.setException (FutureTask.java:252)
  at java.util.concurrent.FutureTask.run (FutureTask.java:271)
  at android.os.AsyncTask$SerialExecutor$1.run (AsyncTask.java:305)
  at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1167)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:641)
  at java.lang.Thread.run (Thread.java:923)
Caused by: java.lang.IllegalStateException: 
  at org.sufficientlysecure.keychain.securitytoken.AutoValue_KdfParameters$Builder.build (AutoValue_KdfParameters.java:21)
  at org.sufficientlysecure.keychain.securitytoken.KdfParameters.fromKdfDo (KdfParameters.java:2)
  at org.sufficientlysecure.keychain.securitytoken.SecurityTokenConnection.retrieveKdfDo (SecurityTokenConnection.java:7)
  at org.sufficientlysecure.keychain.securitytoken.SecurityTokenConnection.calculateKdfIfNecessary (SecurityTokenConnection.java:2)
  at org.sufficientlysecure.keychain.securitytoken.SecurityTokenConnection.verifyPinForOther (SecurityTokenConnection.java:4)
  at org.sufficientlysecure.keychain.securitytoken.operations.PsoDecryptTokenOp.verifyAndDecryptSessionKey (PsoDecryptTokenOp.java:1)
  at org.sufficientlysecure.keychain.ui.SecurityTokenOperationActivity.doSecurityTokenInBackground (SecurityTokenOperationActivity.java:57)
  at org.sufficientlysecure.keychain.ui.base.BaseSecurityTokenActivity.handleSecurityToken (BaseSecurityTokenActivity.java:1)
  at org.sufficientlysecure.keychain.ui.base.BaseSecurityTokenActivity$1.doInBackground (BaseSecurityTokenActivity.java:3)
  at org.sufficientlysecure.keychain.ui.base.BaseSecurityTokenActivity$1.doInBackground (BaseSecurityTokenActivity.java:1)
  at android.os.AsyncTask$3.call (AsyncTask.java:394)
  at java.util.concurrent.FutureTask.run (FutureTask.java:266)
```

(despite how this looks btw, version 5.6 still fixed a lot of crashes for folks. we cut the number of crash reports by half or so :)

**Comment by hagau** [comment]
@Valodim thanks, that should help a lot

**Comment by hagau** [comment]
Ok, i found something: when parsing the KDF-DO, I forgot a `break` statement, thus recognizing a token with KDF disabled (but with KDF ability) as a token with KDF enabled, resulting in the Autovalue error upon parsing the KDF-DO.

My Yubikey 5 NFC for testing with KDF disabled actually did not have KDF ability, since it's firmware version 5.1.2, and I forgot testing with the newer token with KDF ability disabled.
Fixed with #2652 

Sorry, everyone :flushed:

**Comment by tigernero79** [comment]
when will the update come out on the play store?  in fact I have yubico 5 nfc fw 5.2.7 with kdf disabled

Thanks for all

**Comment by tigernero79** [comment]
Found update on playstore 5.6.1 but still the same problem

**Comment by Valodim** [comment]
damn, so it wasn't that. I'll look for other related crash reports later on.

**Comment by sebdehne** [comment]
Same issue with 5.6.1 for me as well. Sent a crash-rapport to google play.

**Comment by tigernero79** [comment]
this time 1 in 10 copied encryption code even though it crashed.  but problem remains with 5.6.1

**Comment by hagau** [comment]
ok, I think I can actually replicate the bug now...

**Comment by hagau** [comment]
Ok, this should now be definitely fixed with #2653 :slightly_smiling_face:

**Comment by tigernero79** [comment]
when will the last modification be public on the play store?  or do I have to subscribe to the beta channel to get it?  will it be 5.6.2 or will it remain version 5.6.1?

**Comment by duhruh** [comment]
running into this myself, not sure how to get the updated version

**Comment by kensmith** [comment]
I'm having this issue as well. OpenKeychain 5.6.1 from Google Play Store. Pixel 5. Yubikey 5 NFC.

**Comment by TraceM171** [comment]
Same problem here, I don't know how to get the updated version.

**Comment by elijahmathews** [comment]
I assume 5.6.2 is what will be the new update, but 5.6.1 is still broken for me. Can't decrypt using my `ed25519` key on either my Yubikey 5 NFC or 5C NFC, both via NFC and USB.

**Comment by hagau** [comment]
OpenKeychain 5.6.2 is available in the Google Play Store, at least on my devices.

**Comment by sebdehne** [comment]
Now it works for me with 5.6.2 using Yubikey 5C NFC. Thanks

**Comment by tigernero79** [comment]
Yes

**Comment by a4b8** [comment]
Thank you for fixing and taking care!

**Comment by kensmith** [comment]
This is fixed for me with 5.6.2. Thank you! Incredible turnaround time!

**Comment by elijahmathews** [comment]
Can confirm it's fixed on 5.6.2. I still get #2590 but it does eventually work.
